cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsEmitter.java

```diff
@@ -33,35 +39,88 @@
   private static final Logger LOG = LoggerFactory.getLogger(AggregatedMetricsEmitter.class);
 
   private final Map<String, String> tags;
```

Comment:
Some logic is wrong. You don't have to keep map of tags in here, as you already have them in a cache entry in a key.
So what you need is
``` java
Cache<Map<String, String>, Cache<String, Measure>>
```

I.e. tags->metric-name->metric